### DISCUSSION TITLE
Update context with assign from a activity

### DISCUSSION THREAD

**Discussion by DirkWolthuis** [statelyai/xstate#1278] [General] [answered]
Is it possible to update the context of the machine from a activity. Like a activity that updates the context with the current time, once a second. When using the assign function inside the activity, it does not update context. I think assign only works in a action context right? 

```javascript
{
    activities: {
      timeIncrement: (context, activity) => {
        // Start the beeping activity
        const timeInterval = setInterval(
          () =>
            assign({
              time: () => new Date(),
            }),
          1000
        );

        // Return a function that stops the beeping activity
        return () => clearInterval(timeInterval);
      },
    },
  }
```

**Comment by Andarist** [comment, marked as answer]
Yes, assign is an action creator - it just returns an object describing what should happen when interpreter sees it. You cant just update the context at arbitrary times like you would like to. Statecharts prefer modelling explicit behaviors and logic so u need to wire this with a parent using events communication.

**Reply by davidkpiano** [reply]
Right - think of activities as "fire-and-forget" with a lifetime defined by the state that it's in.

You're better off modeling this as an invoked callback, and responding to events on the interval:

```js
invoke: {
  src: () => (sendBack) => {
    const timeInterval = setInterval(
      () => sendBack("time.change")
    }, 1000);

    // ... (cleanup)
  },
},
on: {
  'time.change': { actions: assign({ time: () => new Date() }) }
}
```

**Reply by DirkWolthuis** [reply]
Thank you @davidkpiano and @Andarist !

**Comment by DirkWolthuis** [comment]
Ok, I get that. And is there a way to send a event from an activity without
access to the interperted machine?

Op vr 3 jul. 2020 om 17:18 schreef Mateusz Burzyński <
notifications@github.com>

**Reply by Andarist** [reply]
You might want to look into invoked callbacks - activities can be implemented using them, but activities can't communicate with a parent, whereas invoked callbacks can.
